File: AlgorithmsProject/src/in/algorithms/coin/CoinProblem.scala
package in.algorithms.coin

object CoinProblem {
	
  def countNumberOfWays(amount : Integer, denominations : List[Int]) : Int = {
    if(amount == 0) 1
    else if (amount < 0) 0
    else if(denominations.isEmpty) 0
    
    else countNumberOfWays(amount, denominations.tail) + countNumberOfWays(amount - denominations.head, denominations)
  
  }
  
  def main(args: Array[String]) {
	  println("2 with [1,2] is " + countNumberOfWays(2,List(1,2)))
	  println("5 with [1,2] is " + countNumberOfWays(5,List(1,2)))
	  println("5 with [1,2,3] is " + countNumberOfWays(5,List(1,2,3)))
}
}

File: AlgorithmsProject/src/in/algorithms/coin/CoinProblem.java
package in.algorithms.coin;

import java.util.List;

public class CoinProblem {
    public static int countNumberOfWays(int target, List<Integer> coins) {
        if (target == 0) return 1;
        if (target < 0 || coins == null || coins.isEmpty()) return 0;

        int[] dp = new int[target + 1];
        dp[0] = 1;

        for (int coin : coins) {
            for (int i = coin; i <= target; i++) {
                dp[i] += dp[i - coin];
            }
        }
        return dp[target];
    }

    public static int minCoins(int target, List<Integer> coins) {
        if (target == 0) return 0;
        int[] dp = new int[target + 1];
        java.util.Arrays.fill(dp, target + 1);
        dp[0] = 0;

        for (int i = 1; i <= target; i++) {
            for (int coin : coins) {
                if (coin <= i) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[target] > target ? -1 : dp[target];
    }
}
